tests/test_generate_human_tissue_report.py

```diff
@@ -0,0 +1,56 @@
+from unittest.mock import patch
+from datetime import date
+
+from scripts.generate_human_tissue_report import GenerateHumanTissueReport
+from tests.test_common import TestEPP, FakeEntitiesMaker, NamedMock
+
+
+class TestGenerateHumanTissueReport(TestEPP):
+    # today's date required for file name and email
+    todays_date = str(date.today())
+    # csv report needs to be accessible throughout object
+    report_path = 'Human_Tissue_Report_' + todays_date + '.xlsx'
+
+    #fake udfs required for LIMS get call
+    project_udfs = {'REC#':'EthicNumber1', 'Organisation providing ethical consent approval':'Spectra'}
+    sample_udfs1 = {'Freezer':'FREEZER1', 'Shelf':'SHELF1'}
+    sample_udfs2 = {'Freezer': 'Sample Destroyed', 'Shelf': 'Sample Destroyed'}
+
+    #wmock the return value of the function get_human_artifacts that calls lims
+    mock_submitted_samples = [NamedMock(real_name='Sample1',
+                                       project=NamedMock(real_name='X19999', udf=project_udfs),
+                                       udf=sample_udfs1),
+                              NamedMock(real_name='Sample2',
+                                        project=NamedMock(real_name='X19999', udf=project_udfs),
+                                        udf=sample_udfs2)
+                              ]
+    #patch get human artifacts, patch email send and patch saving of report
+    patched_get_human_artifacts = patch('scripts.generate_human_tissue_report.get_human_artifacts', return_value=mock_submitted_samples)
+    patch_email = patch('egcg_core.notifications.email.send_email')
+    patch_workbook_save = patch('openpyxl.workbook.workbook.Workbook.save')
+
+    def setUp(self):
+        self.epp = GenerateHumanTissueReport(self.default_argv)
+
+    def test_generate_report_and_send_email(self):
```

Comment:
This test that the email has been sent but does not check the content of the attachment.
You should let the file be saved to disk and check the md5 (if it is deterministic).
You could also patch `Workbook` `__setitem__` (which is used by the `ws['item']`) to capture what was entered in each cell and test that.

Reply:
I have updated to test the md5sum of the file. I have also introduced a test that tests the last instance of __setitem__. I wasn't sure how to directly test all instances of this.

Reply:
There are a couple of ways of doing it:

```python
# assert_any_call - assert that something has been called a certain way at some point
mocked_thing.assert_any_call('this', 'that')
mocked_thing.assert_any_call('this', 'that', other='another')

# assert_has_calls - assert that something has only had these exact calls
from unittest.mock import call
mocked_thing.assert_has_calls(
    [
        call('this', 'that'),
        call('this', 'that', other='another')
    ]
)
```